### PR TITLE
Support collection of remote logging data in testify

### PR DESCRIPTION
**Pull request by joe-chacko**

### CODE REVIEW

**Review by habiblawal1 — DISMISSED** [review]

File: testify/src/main/java/testify/annotation/logging/LogRecorder.java [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+    }
+
+    static void dispatchReply(Bus bus, String s) {
+        int requestId = bus.get(REQUEST_ID);
+        bus.put(String.format(RECEIPT_FORMAT, requestId), s);
+    }
+
+    private static int getNextRequestId(Bus dedicatedBus) {
+        synchronized (dedicatedBus) {
+            final int requestId;
```
Comment (ngmr) [inline testify/src/main/java/testify/annotation/logging/LogRecorder.java:185]: declaration and initialization of 'requestId' could be combined on one line

File: testify/src/main/java/testify/annotation/logging/LogPublisher.java [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+                    .sorted()
+                    .forEach(out::println);
+
+            out.printf("%c%1$c%1$cEND LOG REPLAY [%s] %1$c%1$c%1$c%n", flag, displayName);
+            out.flush();
+        }
+        return this;
+    }
+
+    public synchronized void close() {
```
Comment (ngmr) [inline testify/src/main/java/testify/annotation/logging/LogPublisher.java:127]: Is this close idempotent ?
Should LogPublisher describe itself as AutoCloseable ?

File: testify/src/main/java/testify/annotation/logging/LogRecorder.java [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+                return sw.toString();
+            }
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    static void close(Bus dedicatedBus) { dispatchRequestAndWaitForReply(dedicatedBus, CLOSE); }
+
+    private synchronized void close() {
```
Comment (ngmr) [inline testify/src/main/java/testify/annotation/logging/LogRecorder.java:353]: Is this close() idempotent ?
Should LogRecorder declare itself as AutoCloseable ?

File: testify/src/main/java/testify/annotation/logging/LoggingExtension.java [hunk 4]
```diff
@@ -25,21 +25,77 @@
...
+    public void beforeTestExecution(ExtensionContext ctx) { startLogging(ctx); }
+    public void afterTestExecution(ExtensionContext ctx) { endLogging(ctx); }
+    public void afterAll(ExtensionContext ctx) { endLogging(ctx); }
     public void handleTestExecutionException(ExtensionContext ctx, Throwable throwable) throws Throwable {
-        this.somethingWentWrong(throwable);
+        if (null != logPublisher) logPublisher.somethingWentWrong(throwable);
         throw throwable; // rethrow or tests won't fail
     }
+
     public void close() throws Throwable {
```
[the remark this answers is not in the document]
Comment (joe-chacko) [reply testify/src/main/java/testify/annotation/logging/LoggingExtension.java:77]: Yes — both the methods it calls are idempotent.

File: testify/src/main/java/testify/annotation/logging/LogRecorder.java (hunk 1 above)
Comment (joe-chacko) [reply testify/src/main/java/testify/annotation/logging/LogRecorder.java:185]: Changed.

File: testify/src/main/java/testify/annotation/logging/LogRecorder.java (hunk 3 above)
Comment (joe-chacko) [reply testify/src/main/java/testify/annotation/logging/LogRecorder.java:353]: I tried AutoCloseable, but IntelliJ complained when it was created outside of a try-with-resources statement.

File: testify/src/main/java/testify/annotation/logging/LogPublisher.java (hunk 2 above)
Comment (joe-chacko) [reply testify/src/main/java/testify/annotation/logging/LogPublisher.java:127]: I tried AutoCloseable, but IntelliJ then complained if I assigned it outside a try-with-resources parameter list.

**Review by joe-chacko — COMMENTED** [review]
Responded to review comments and made requested changes.

**Review by habiblawal1 — APPROVED** [review]

**Pull request merged**